### PR TITLE
ADT75 driver + demo project

### PR DESCRIPTION
**Pull request by CiprianRegus**

### CODE REVIEW

File: drivers/temperature/adt75/adt75.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+/******************************************************************************/
+
+/**
+ * @brief Get a register's size in bytes
+ * @param addr - The register's address
+ * @return The register's size
+ */
+static int adt75_reg_size(uint32_t addr)
+{
+	if (addr == ADT75_CONF_REG)
```
Comment (amiclaus) [inline drivers/temperature/adt75/adt75.c:61]: You can do this as a macro, and drop this extra function: https://github.com/torvalds/linux/blob/b7b275e60bcd5f89771e865a8239325f86d9927d/drivers/iio/dac/ad7293.c#L21-L33 .
Or just do the check inline.

File: include/no_os_units.h [hunk 2]
```diff
@@ -65,4 +65,6 @@
 #define MICROWATT_PER_MILLIWATT	1000UL
 #define MICROWATT_PER_WATT	1000000UL
 
+#define MILLIDEGREE_PER_DEGREE  1000UL
```
Comment (amiclaus) [inline include/no_os_units.h:68]: move to separate commit

File: projects/adt75/src/examples/dummy/dummy_example.c [hunk 3]
```diff
@@ -0,0 +1,91 @@
+/***************************************************************************//**
```
Comment (danmois) [inline projects/adt75/src/examples/dummy/dummy_example.c:1]: There was a discussion regarding the switch from renaming the example from "DUMMY" to "BASIC".

File: projects/adt75/src/examples/dummy/dummy_example.c [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		pr_info("%c",27);
+		pr_info("%c",'[');
+		pr_info("%c",'2');
+		pr_info("%c",'J');
+
+		ret = adt75_get_single_temp(adt75_desc, &val);
+		if (ret)
+			goto error;
+
+		printf("Temperature: %.03f C\n", (double)val / 1000);
```
Comment (danmois) [inline projects/adt75/src/examples/dummy/dummy_example.c:82]: Maybe using pr_info instead of printf would be better.

**Comment by CiprianRegus** [comment]
- Removed the `adt75_reg_size()` function, and computed the register size inline.
- Moved the addition of the `MILLIDEGREE_PER_DEGREE` to another commit.
- Replaced the `printf()` calls with `pr_info()` in the example project.
- Fixed the warnings in the IIO driver.
- Removed the channels update function from the IIO driver, since there is only 1 channel (which hardware wise, it's always enabled).

File: drivers/temperature/adt75/adt75.c [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+/******************************************************************************/
+
+/**
+ * @brief Read the raw value of a register.
+ * @param desc - The device structure.
+ * @param addr - The register's address.
+ * @param val - The value to be read.
+ * @return 0 in case of success, -EINVAL otherwise
```
Comment (amiclaus) [inline drivers/temperature/adt75/adt75.c:59]: nit: are you sure the negative return code will always be EINVAL?

File: drivers/temperature/adt75/adt75.c [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+ */
+int adt75_reg_read(struct adt75_desc *desc, uint32_t addr, uint16_t *val)
+{
+	int reg_size;
+	int ret;
+
+	reg_size = (addr == ADT75_CONF_REG)? 1 : 2;
+
+	desc->buff[0] = no_os_field_get(ADT75_DATA_REG_MASK, addr);
```
Comment (amiclaus) [inline drivers/temperature/adt75/adt75.c:68]: can we reduce the scope of `buff` to the read/write function and drop it from the device structure?

File: drivers/temperature/adt75/adt75.c [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	ret = adt75_reg_read(desc, ADT75_TEMP_VALUE_REG, &reg_val);
+	if (ret)
+		return ret;
+
+	/** Wait for a new conversion to finish */
+	no_os_mdelay(ADT75_CONV_DELAY_MS);
+
+	reg_val = no_os_field_get(ADT75_TEMP_MASK, reg_val);
+	*val = no_os_sign_extend32(reg_val, ADT75_SIGN_BIT);
+	*val *= MILLIDEGREE_PER_DEGREE;
```
Comment (amiclaus) [inline drivers/temperature/adt75/adt75.c:125]: nit: can reduce here the number of lines: `*val = *val * MILLIDEGREE_PER_DEGREE / ADT75_TEMP_DIV;`

File: drivers/temperature/adt75/adt75.h [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+/** Configuration register fields */
+#define ADT75_SHUTDOWN_MASK		NO_OS_BIT(0)
+#define ADT75_COMP_INT_MASK		NO_OS_BIT(1)
+#define ADT75_ALERT_POL_MASK	NO_OS_BIT(2)
+#define ADT75_FAULT_QUEUE_MASK	NO_OS_GENMASK(4, 3)
+#define ADT75_ONESHOT_MASK		NO_OS_BIT(5)
+
+#define ADT75_HYST_MASK			NO_OS_GENMASK(15, 4)
+#define ADT75_OVER_TEMP_MASK	NO_OS_GENMA.SK(15, 4)
```
Comment (amiclaus) [inline drivers/temperature/adt75/adt75.h:70]: typo here `NO_OS_GENMASK(15, 4)`

File: drivers/temperature/adt75/adt75.h [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#ifndef _ADT75_H
+#define _ADT75_H
+
+#include "stdint.h"
```
Comment (buha) [inline drivers/temperature/adt75/adt75.h:42]: use <> here

File: projects/adt75/Makefile [hunk 10]
```diff
@@ -0,0 +1,9 @@
+# Select the example you want to enable by chossing y for enabling and n for disabling
+DUMMY_EXAMPLE = y
```
Comment (buha) [inline projects/adt75/Makefile:2]: yeah, as ohers have said, maybe rename DUMMY to BASIC

File: projects/adt75/src/examples/dummy/dummy_example.c [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+	int32_t val;
+	struct adt75_desc *adt75_desc;
+
+	ret = adt75_init(&adt75_desc, &adt75_ip);
+	if (ret)
+		return ret;
+
+	while (1) {
+		/* Clear the screen. */
+		pr_info("%c",27);
```
Comment (danmois) [inline projects/adt75/src/examples/dummy/dummy_example.c:73]: Maybe printf should be used here. Problems could arise when using timestamps for pr_info.

File: projects/adt75/src/examples/iio_example/iio_example.c [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#define DATA_BUFFER_SIZE 400
+
+/******************************************************************************/
+/************************ Variable Declarations ******************************/
+/******************************************************************************/
+uint8_t iio_data_buffer[DATA_BUFFER_SIZE * sizeof(int32_t)];
```
Comment (rbolboac) [inline projects/adt75/src/examples/iio_example/iio_example.c:56]: taking into consideration that the scan type is int16_t, it would make sense to allocate a buffer with the same size multiplier

File: projects/adt75/src/examples/iio_example/iio_example.c [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+ *               execute continuously function iio_app_run and will not return.
+*******************************************************************************/
+int iio_example_main()
+{
+	int ret;
+	struct adt75_iio_desc *adt75_iio_desc;
+	struct adt75_iio_init_param adt75_iio_ip;
+	struct iio_data_buffer buff = {
+		.buff = (void *)iio_data_buffer,
+		.size = DATA_BUFFER_SIZE * sizeof(uint32_t)
```
Comment (rbolboac) [inline projects/adt75/src/examples/iio_example/iio_example.c:74]: taking into consideration that the scan type is int16_t, it would make sense to obtain the size of the buffer with the same type size multiplier

File: drivers/temperature/adt75/adt75.c [hunk 14]
```diff
@@ -0,0 +1,179 @@
+/***************************************************************************//**
+ *   @file   adt75.c
+ *   @brief  Source file of ADT75 Driver.
+ *   @author Ciprian Regus (ciprian.regus@analog.com)
+********************************************************************************
+ * Copyright 2022(c) Analog Devices, Inc.
```
Comment (rbolboac) [inline drivers/temperature/adt75/adt75.c:6]: even though this is an older pull request, it would be nice to change the year to 2023 in all files

File: drivers/temperature/adt75/adt75.h [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+/** x is set based on the value of A2, A1, A0 pins */
+#define ADT75_ADDR(x)		(0x48 + (x))
+#define ADT75_CONV_DELAY_MS     40
+#define ADT75_FRAME_SIZE	3
+
+#define ADT75_TEMP_VALUE_REG    0x0
+#define ADT75_CONF_REG          0x1
+#define ADT75_HYST_REG          0x2
+#define ADT75_OS_REG            0x3
+#define ADT75_ONE_SHOT_REG      0x4s
```
Comment (amiclaus) [inline drivers/temperature/adt75/adt75.h:54]: was `s` added on purpose?

**Comment by CiprianRegus** [comment]
- Changed size of the IIO buffer, according to the sample size.
- 2022 -> 2023
- removed the `s` character.

**Review by amiclaus — APPROVED** [review]

**Review by rbolboac — APPROVED** [review]
